### ISSUE TITLE
Parallel Task Graph

### ISSUE THREAD

**Issue by sholloway** [sholloway/agents-playground#164]
Create a new task graph runner that runs tasks in parallel. 

### Considerations

- The OOTB [graphlib](https://docs.python.org/3/library/graphlib.html) package might help organize the task graph. It is designed to sort an adjacency list (represented as a dict) for parallel processing. It can also detect cycles.

### Serialization 
- cloudpickle: Makes it possible to serialize Python constructs not supported by the default pickle module from the Python standard library.
- An option to pass data (e.g. resources) between processes is to use the file system. Dask leverages several packages to help with this. fsspec for mounting filesystems and partd for an interprocess key/value store. 
- Consider compressing data between processes. The package partd has some capabilities with this.
- Could [Apache Arrow](https://arrow.apache.org/) be used for shoving data around?

### Async and Parallelization 
- A thought is to leverage sockets to communicate across processes. This could be for messaging (as opposed to moving data back and forth). If sockets make sense, one option is to leverage 0MQ (pyzmq package). tornado is another option.
- From a design perspective, perhaps think about how the main process and worker process could function like a CoreOS cluster. Having an orchestrator and workers that sync over an etcd shared memory pool. The package opencensus might have some ideas on how to monitor this type of design.
-
